java/arcs/core/entity/CollectionHandle.kt

```diff
@@ -119,16 +119,24 @@ class CollectionHandle<T : Storable, R : Referencable>(
             "Entity must have an ID before it can be referenced."
         }
 
-        adaptValues(storageProxy.getParticleViewUnsafe()).let { data ->
-            data.firstOrNull()?.let {
-                require(it is Entity) {
-                    "Handle must contain Entity-typed elements in order to create references."
+        requireNotNull(
+            /**
+             * [referencableToStorable] deserialization is expensive.
+             * Traverse and match in order is much cheaper than deserialize all then look up.
+             *
+             * As [CrdtSet.DataImpl.values] is in [MutableMap] type which uses [LinkedHashMap] in
+             * kotlin, [CrdtSet] model merging logic is the newest data gets inserted at the head
+             * which benefits the traversal to expedites common use-cases i.e.
+             * store-then-createReference-immediately from time complexity O(N) to O(1).
+             */
+            storageProxy.getParticleViewUnsafe().firstOrNull {
+                when (val maybeEntity = storageAdapter.referencableToStorable(it)) {
+                    is Entity -> !storageAdapter.isExpired(maybeEntity) && entityId == entityId
```

Comment:
We should be able to check the expiration time and compare the entityId without conversion as well, right? So it should only be necessary to convert the actual matched entity?

Reply:
Yes though `StorageAdapter.isExpired` can only accept a `Storable` now.
I thought copying the `isExpired` logic to here to improve another (0-0.5 ms) but it potentially increases the difficulty of future maintenance.
Perhaps I can just overload `isExpired` in StorageAdapter.kt to accept both of `Storable` and `Referencable` :)

Reply:
In that case we could move the timestamp defs into `Storable` interface, and have `Referencable` extend `Storable`.

But I just realized one thing I'm not sure about: are the timestamps of the reference necessarily the same as the timestamps of the backing entity? Maybe this is not actually possible, after all.

Reply:
Looking at the code, it seems right now that we treat them as the same. And this patch passes all tests:  https://github.com/jblebrun/arcs/pull/new/storable-interface

Reply:
But in a chat discussion, @mmandlis confirmed that we should not make this assumption. 

But I think it could still be OK to compare entityId before the conversion?

Reply:
That's also not true, so ignore this entire conversation. :-)

Reply:
I'm silly: These are `Referencable`, not `Reference`... so actually it should be fine to compare the entity ID and timestamp before conversion!

Reply:
:)
yep, we can compare timestamp and entity id before conversion/deserialization.
The StorageAdapter.isExpire() controls the expiration logic i.e. time.currentTimeMillis may change some day to use epoch or other representations, so I didn't want to copy it here for better future maintenance.

I'll add overload functions to StorageAdapter.kt to accept both of Storable and Referenable.
Then modify this PR accordingly to do the comparison before conversion/deserialization.

Reply:
Just posted to chat, but why do we need to do this check? it doesn't matter if the entity is in the collection, we just need its id. Can we try remove this block entirely and run tests?

Reply:
I tried to remove the check-existence block but failed a few tests like HandleAdapterTest, HandleManagerCloseTest, etc.

It makes sense to create-reference-after-check-existence from semantics.
The point is to minimize the check-existence overhead by reducing or completely getting rid of expensive referencableToStorable.

Reply:
I've got rid of all `referencableToStorable` in the new uploaded commit.
Now just use `Referenable` object to compare entity id and check expiration.
The `createReference` runs quite fast now :-)